samples/cellular/modem_shell/src/sms/sms.c

```diff
@@ -99,27 +100,49 @@ int sms_unregister(void)
 }
 
 /* Function name is not sms_send() because it's reserved by SMS library. */
-int sms_send_msg(char *number, char *text)
+int sms_send_msg(char *number, char *data, enum sms_data_type type)
 {
 	int ret;
+	uint16_t data_hex_len;
+	uint8_t *data_hex;
 
 	if (number == NULL || strlen(number) == 0) {
 		mosh_error("Number not given");
 		return -EINVAL;
 	}
-	if (text == NULL || strlen(text) == 0) {
+	if (data == NULL || strlen(data) == 0) {
 		mosh_error("Text not given");
 		return -EINVAL;
 	}
 
-	mosh_print("Sending SMS to number=%s, text='%s'", number, text);
+	mosh_print("Sending SMS to number=%s, text='%s'", number, data);
+	if (type == SMS_DATA_TYPE_GSM7BIT) {
+		mosh_print("Input text is in GSM 7bit format");
+	}
 
 	ret = sms_register();
 	if (ret != 0) {
 		return ret;
 	}
 
-	ret = sms_send_text(number, text);
+	if (type == SMS_DATA_TYPE_GSM7BIT) {
+		/* Process data to be sent if it's in hex format */
+		data_hex_len = strlen(data) / 2;
```

Comment:
If the data length is not divisible by 2, the buffer will be too small. That will cause `str_hex_to_bytes()` to not do the conversion and the sent data will be 0 bytes.

If input data always needs to be divisible by 2, it would be good to have check for it and print an error. Otherwise, the length should be calculated like in `hex2bin()`:
```
data_len = strlen(data);
data_hex_len = data_len / 2 + data_len % 2;
```
And one more thing, aren't the names `data_hex` and `data_hex_len` misleading? Input data is hex, but the output is binary, right?

Reply:
Yep, I've seen empty messages when hex data is not divisible by 2. I'll add a check that the length is divisible by 2.

I tried how it would work with hex strings such as `616263646` and `hex2bin()` adds `0` in front of the data, which will change the content of the SMS message look completely different and hence confusing if you accidentally miss one number from the hex data.

`data_hex*` renamed to `data_bin*`.

Reply:
> Yep, I've seen empty messages when hex data is not divisible by 2. I'll add a check that the length is divisible by 2.

Is there a test for this, at least I didn't spot it?

Reply:
Not yet, I'll check the branch coverage data today and try to fill the gaps, including this one.

Reply:
Coverage data for the library unit tests still looks good. This change is in MOSH so no tests for that.

Reply:
> Coverage data for the library unit tests still looks good. This change is in MOSH so no tests for that.

Of course, my bad.